src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -572,23 +573,14 @@ void ObjectSynchronizer::exit(oop object, BasicLock* lock, JavaThread* current)
     markWord mark = object->mark();
     if (LockingMode == LM_LIGHTWEIGHT) {
       // Fast-locking does not use the 'lock' argument.
-      if (mark.is_fast_locked()) {
-        markWord unlocked_mark = mark.set_unlocked();
-        markWord old_mark = object->cas_set_mark(unlocked_mark, mark);
-        if (old_mark != mark) {
-          // Another thread won the CAS, it must have inflated the monitor.
-          // It can only have installed an anonymously locked monitor at this point.
-          // Fetch that monitor, set owner correctly to this thread, and
-          // exit it (allowing waiting threads to enter).
```

Comment:
I don't understand why the anonymous owner case is no longer being checked. ??

Reply:
The condition does now check for a successful CAS, not the unsuccessful one. If it was successful then there is no monitor, thus no anonymous owner.

If the CAS failed and the mark word is no longer fast locked. It must be inflated. So we fallthrough down to the inflated case.

`ObjectSynchronizer::inflate` correctly handles fixing the owner.

Reply:
I see - thanks. It is hard to see where the code goes to when the CAS fails.

Reply:
In the old code, I think if there was a hash code installation racing with this
monitor exit by the thread owner, the hash code installer thread would have
inflated the monitor which would have resulted in the anonymous owner value
also being set. So we could have had two different inflating thread racers...